File: Packages/com.github.homuler.mediapipe/Runtime/Scripts/Framework/OutputStreamPoller.cs
// Copyright (c) 2023 homuler
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

using System;

namespace Mediapipe
{
  public class OutputStreamPoller<T> : MpResourceHandle
  {
    public OutputStreamPoller(IntPtr ptr) : base(ptr) { }

    protected override void DeleteMpPtr()
    {
      UnsafeNativeMethods.mp_OutputStreamPoller__delete(ptr);
    }

    public bool Next(Packet<T> packet)
    {
      UnsafeNativeMethods.mp_OutputStreamPoller__Next_Ppacket(mpPtr, packet.mpPtr, out var result).Assert();

      GC.KeepAlive(this);
      return result;
    }

    public void Reset()
    {
      UnsafeNativeMethods.mp_OutputStreamPoller__Reset(mpPtr).Assert();
    }

    public void SetMaxQueueSize(int queueSize)
    {
      UnsafeNativeMethods.mp_OutputStreamPoller__SetMaxQueueSize(mpPtr, queueSize).Assert();
    }

    public int QueueSize()
    {
      UnsafeNativeMethods.mp_OutputStreamPoller__QueueSize(mpPtr, out var result).Assert();

      GC.KeepAlive(this);
      return result;
    }
  }
}

File: mediapipe_api/framework/output_stream_poller.cc
// Copyright (c) 2021 homuler
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

#include "mediapipe_api/framework/output_stream_poller.h"

void mp_OutputStreamPoller__delete(mediapipe::OutputStreamPoller* poller) { delete poller; }

MpReturnCode mp_OutputStreamPoller__Reset(mediapipe::OutputStreamPoller* poller) {
  TRY_ALL
    poller->Reset();
    RETURN_CODE(MpReturnCode::Success);
  CATCH_ALL
}

MpReturnCode mp_OutputStreamPoller__Next_Ppacket(mediapipe::OutputStreamPoller* poller, mediapipe::Packet* packet, bool* result_out) {
  TRY_ALL
    *result_out = poller->Next(packet);
    RETURN_CODE(MpReturnCode::Success);
  CATCH_ALL
}

MpReturnCode mp_OutputStreamPoller__SetMaxQueueSize(mediapipe::OutputStreamPoller* poller, int queue_size) {
  TRY_ALL
    poller->SetMaxQueueSize(queue_size);
    RETURN_CODE(MpReturnCode::Success);
  CATCH_ALL
}

MpReturnCode mp_OutputStreamPoller__QueueSize(mediapipe::OutputStreamPoller* poller, int* queue_size_out) {
  TRY_ALL
    *queue_size_out = poller->QueueSize();
    RETURN_CODE(MpReturnCode::Success);
  CATCH_ALL
}
